### ISSUE TITLE
Less sensitive check for available GPU memory

### ISSUE THREAD

**Issue by Masaaki-75** [CERN/TIGRE#574]
## TLDR
Request to modify the GPU usage check in TIGRE to allow functions to run with sufficient memory available, instead of requiring the GPU to be completely idle.

## Verbose
I just finished installing TIGRE python version on Ubuntu OS, and started to test the example in `Python/example.py`. I modified the code a little to specify the GPU card (by adding `os.environ['CUDA_VISIBLE_DEVICES'] = '1'` right after `import os`).

At first, there was another program that was running with 16/24G of my GPU card (and the GPU utilization is also low, meaning i have 8G free memory to run other programs), but i got the stdout message telling me to free the GPU. So i closed that old program and freed 16G memory usage on the GPU card 1, and the example code ran successfully.

When i checked the GPU usage of this example code, it showed < 400M memory was used. However, the code didn't start even when i have 8G memory free. So I was hoping if you could make TIGRE's detection of available GPU memory less sensitive, so the related code can be run once there is available memory (but not requiring a totally idle GPU).

## Expected Behavior
It would be more practical if the functions in TIGRE checked for **sufficient available GPU memory rather than requiring complete GPU idleness**. Functions should execute as long as there is enough memory to accommodate them, regardless of other ongoing processes.

## Actual Behavior
Currently, if any program is running on the GPU, even if it uses minimal memory and leaves ample memory free (e.g., over 1GB), the relevant functions in TIGRE refuse to execute and return the following error:
```bash
0: NVIDIA GeForce RTX 4090
{'name': 'NVIDIA GeForce RTX 4090', 'devices': [0]}
../Common/CUDA/TIGRE_common.cpp (14): tvDenoise:tvdenoising:GPU One (or more) of your GPUs is being heavily used by another program (possibly graphics-based).
 Free the GPU to run TIGRE
```

## Specifications 

  - Python version: 3.9
  - OS: Ubuntu 18.04.6
  - CUDA version: 12.2
  - GPU: NVIDIA GeForce RTX 4090

**Comment by AnderBiguri** [comment]
Hi!
You can change this in the following piece of code

https://github.com/CERN/TIGRE/blob/729f146316c0d214b2f00cfbdc1111490032e35e/Common/CUDA/Siddon_projection.cu#L855

You'll also need to do it in the voxelbackprojection file. Currently it's set at requiring 95% of the memory free.

Now, this is not recomended. While TIGRE may not need all the memory, it will take 100% of the compute time, so launching it at the same time as another process is using the gpu may make the total time of running them in parallel slower than the serial version.

**Comment by Masaaki-75** [comment]
Thanks for the prompt reply!

BTW, since i already install TIGRE in my conda, is it possible to locate and modify this parameter in the compiled code? (i guess it should be at somewhere in the conda environment like `.../miniconda3/envs/tigre/lib/python3.9/site-packages/tigre/` but i failed to locate it). Cuz if not, then i'll have to reinstall the modified version.

**Comment by AnderBiguri** [comment]
There is no con da distribution of Tigre (ie conda install tigre) , so you must have downloaded via git clone. so just go to the files downloaded, modify them, and pip install again, as per instructions.

Note that you may need to change other cu files too, same code.

**Comment by Masaaki-75** [comment]
I tried modifying the number 0.95 to 0.01 in the following files that have definition for `checkFreeMemory`, and then re-create an virtual environment and re-compile TIGRE:
- POCS_TV.cu
- POCS_TV2.cu
- ray_interpolated_projection.cu
- Siddon_projection.cu
- tvdenoising.cu
- voxel_backprojection.cu
- voxel_backprojection2.cu

So i got one dummy program using GPU running on my card, and i can control the memory usage of this program. I found that with the above modification, `example.py` still complains about the GPU even when there're >8G memory free to use (the error message is like before). 
Actually, `example.py` didn't run successfully until i lowered GPU memory usage of my dummy program to around 10G (so there're 24-10=14G free to use). 

Is it normal? (I thought 0.01 is a quite small number). Or did i miss something?

The success message is like:
```
0: NVIDIA GeForce RTX 4090
{'name': 'NVIDIA GeForce RTX 4090', 'devices': [0]}
OSSART algorithm in progress.
Estimated time until completion : 00:01:42
RMSE fdk:
376.5443928704311
RMSE ossart
212.95980004117254
```

**Comment by AnderBiguri** [comment]
Indeed, that should not happen, I'd expect it to never complain if you lowered that much. You can also just comment out completely the if condition, and let the code naturally crash if it happens to try to use too much memory. The only difference is that the message will be different and less clear, otherwise this function doesn't do much else.

**Comment by Masaaki-75** [comment]
Thanks for the hints! Instead of modifying the factor, one should modify the `if` condition:
https://github.com/CERN/TIGRE/blob/729f146316c0d214b2f00cfbdc1111490032e35e/Common/CUDA/Siddon_projection.cu#L848

I change it from `if(memfree<memtotal/2)` to `if(memfree<memtotal/10)` and can run `example.py` smoothly in an already partially occupied GPU!
